wip for BlockchainBackend

SWvheerden opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
RFC for the block chain refactor. 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
We need to refactor the BlockchainBackend as it depends too much on a keyvalue store

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [ ] I'm merging against the `development` branch.
* [ ] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

CjS77 commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -133,21 +134,28 @@ impl<B: BlockchainBackend> Clone for Validators<B> {
-    /// Commit the transaction given to the backend. If there is an error, the transaction must be rolled back, and
-    /// the error condition returned. On success, every operation in the transaction will have been committed, and
+    /// Adds a block to the orphan database. On success, every operation in the transaction will have been committed,
+    /// and the function will return `Ok(())`.
```
.. and on failure, the state of the DB will be exactly as it was before the call to add_orphan_block

CjS77 replied:
Aspects of the contract you need to address:
* Does add_orphan_block trigger any re-org logic? Or does it simply add the block and leave it there?
* Does it do any validation? Or willit add invalid blocks to the orphan db?
* What if the block is already present as an orphan block?
* What if the block is already present on the main chain?

SWvheerden replied:
I think the first 2 should definitely stay in the database_backend where it currently lives. 
The bottom two I think the check logic should exist in the database_backend, but perhaps we could add safety checks in the db?

CjS77 commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -133,21 +134,28 @@ impl<B: BlockchainBackend> Clone for Validators<B> {
+    /// Adds a block to the orphan database. On success, every operation in the transaction will have been committed,
+    /// and the function will return `Ok(())`.
+    fn add_orphan_block(&mut self, block: Block) -> Result<(), ChainStorageError>;
+    /// This function moves the block from the orphan pool to the main db. On success, every operation in the
```
You're suggesting that the block must have been previously added with `add_orphan_block`? Please be explicit.

On failure, the state must be the same as before the call.

What is the contract regarding:
* What validation is performed?
* What if the blockhash is not present?
* What if the block hash is already present on the main chain?

SWvheerden replied:
The current database_backend does it this way. 
1. add oprhan
2. depending on reorg logic, delete from oprhan and readd to db

I was thinking to perhaps make the database more mimic the backend workings. A block is always first added as orphan then maybe added as block. Never just added as block

CjS77 commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -133,21 +134,28 @@ impl<B: BlockchainBackend> Clone for Validators<B> {
+    /// transaction will have been committed, and the function will return `Ok(())`.
+    fn add_block(&mut self, block_hash: HashOutput) -> Result<(), ChainStorageError>;
+    // rewinds the database to the specified height. On success, every operation in the transaction will have been
+    // committed, and
```
What happens to old blocks and header? Must they be placed in the orphan pool, or is that optional?
If an error ocurrs during rewind, what state is the DB left in?

SWvheerden replied:
This step should just:
1. add block to db.
2. If added, remove from oprhan. 
3. error is returned if this is not successful. Ensure state is reversed.

CjS77 commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -133,21 +134,28 @@ impl<B: BlockchainBackend> Clone for Validators<B> {
-    /// Fetches the merklish root for the MMR tree identified by the key. This function should only fail if there is an
-    /// access or integrity issue with the back end.
-    fn fetch_mmr_root(&self, tree: MmrTree) -> Result<HashOutput, ChainStorageError>;
+    // DELETE? fn fetch_mmr_root(&self, tree: MmrTree) -> Result<HashOutput, ChainStorageError>;
```
I believe we still need this

SWvheerden replied:
This is never called in code, only in unit tests.

CjS77 commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -133,21 +134,28 @@ impl<B: BlockchainBackend> Clone for Validators<B> {
+    fn rewind_to_height(&mut self, height: u64) -> Result<(), ChainStorageError>;
+    /// Commit the transaction given to the backend. This will create a db write transaction. This inly for syncing. On
+    /// success, every operation in the transaction will have been committed, and the function will return `Ok(())`.
+    fn write_sync(&mut self, tx: DbTransaction) -> Result<(), ChainStorageError>;
```
I think since we're going down the road of individual API functions in this trait with explicit contracts, we can think about doing away with write_sync and DbTransaction completely

SWvheerden replied:
I was going down this route as a sync call, as when you download/ sync data, you only ever add data directly to the db as you would have downloaded them. This is more inline with the new pruning sync. You would download the data in packs, like first headers, then kernels etc. There is no more blocks. So rather make a sync function to pipe data directly into the db with almost no extra logic. Thats my thinking behind this function.

CjS77 reviewed (commented):
Right track; we must just be much more explicit and specific about the contract the interface functions are providing

SWvheerden commented:
We should probably take time to think where should the logic live. 

I am currently thinking along the lines of this trait should only care about the data going directly into the db, and thats all. Validation etc is out of bounds for this trait. All calls should be made reversible by the implementation of the trait. So if add_black for example errors it should revere it as if no block was ever added.

sdbondi commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -129,25 +129,34 @@ impl<B: BlockchainBackend> Clone for Validators<B> {
+    fn add_orphan_block(&mut self, block: Block) -> Result<(), ChainStorageError>;
+    /// This function will move a block from the orphan pool to the main chain. It assumes that the block has passed
+    /// full stateful validation
+    fn add_block(&mut self, block_hash: HashOutput) -> Result<(), ChainStorageError>;
```
Maybe `accept_block`?

sdbondi commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -157,42 +166,44 @@ pub trait BlockchainBackend: Send + Sync {
+    //     Self: Sized,
+    //     F: FnMut(Result<(HashOutput, Block), ChainStorageError>);
+    /// returns a list of orphan block headers that are parents to the named hash
+    fn get_parent_orphan_headers(&self, hash: HashOutput) -> Result<Vec<BlockHeader>, ChainStorageError>;
```
Maybe we should keep things consistent and always use `get` or `fetch` - to me, `fetch` implies that some I/O has to happen on each call, whereas `get` doesn't really tell you either way. Since these will likely involve actually fetching the data on each call, I'd go with `fetch`. But I think either is ok.

sdbondi commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -157,42 +166,44 @@ pub trait BlockchainBackend: Send + Sync {
+    /// returns a list of orphan block headers that are parents to the named hash
+    fn get_parent_orphan_headers(&self, hash: HashOutput) -> Result<Vec<BlockHeader>, ChainStorageError>;
+    /// Returns a list of all orphan block headers
+    fn get_all_orphan_headers(&self, hash: HashOutput) -> Result<Vec<BlockHeader>, ChainStorageError>;
```
Just get `get_orphan_headers`? I don't think it makes sense to return multiple headers given a hash?

SWvheerden replied:
thans, thats a copy paste error. Fixed that

sdbondi commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -129,25 +129,34 @@ impl<B: BlockchainBackend> Clone for Validators<B> {
+    fn add_mmr(&mut self, tree: MmrTree, hashes: Vec<HashOutput>) -> Result<(), ChainStorageError>;
     /// Fetch a value from the back end corresponding to the given key. If the value is not found, `get` must return
     /// `Ok(None)`. It should only error if there is an access or integrity issue with the underlying back end.
     fn fetch(&self, key: &DbKey) -> Result<Option<DbValue>, ChainStorageError>;
```
For reference:
```rust
pub enum DbKey {
    Metadata(MetadataKey),
    BlockHeader(u64),
    BlockHash(BlockHash),
    UnspentOutput(HashOutput),
    SpentOutput(HashOutput),
    TransactionKernel(HashOutput),
    OrphanBlock(HashOutput),
}
```

In postgres, you'd probably have to switch between 5 different tables to implement this fetch depending on `DbKey`. Perhaps we should split into `fetch_block(BlockKey::Hash(xxxx))` or `fetch_block(BlockKey::Height(123))` - `fetch_metadata(MetadataKey::ChainHeight)` etc.

sdbondi replied:
Same for `contains`, though we should only add the functions that are used.

sdbondi replied:
Spoken to @SWvheerden the implementation is pretty clean without splitting these up :+1:

SWvheerden committed:
RFC for block chain backend

lmdb changes
get block db transaction
Add rewind to lmdb
add rewind to memory db
refactor of block chain backend code

SWvheerden committed:
merge conflicts of pruning horizon

SWvheerden commented:
Moved this to fork
